TensorImpl: Lazily compute numel and contiguity when symbolic

peterbell10 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #112785
* #112890
* #112689

Currently whenever the sizes or strides are modified for a `TensorImpl` we
eagerly recompute the numel and memory format flags. This is fine for static
shapes as it's all fast C++ code, but for symbolic shapes it runs slow python code.

This instead changes the `SymbolicShapeMeta` object to compute the derived
quantities lazily at the first request. This has the added benefit that we can
now pass assumptions in `empty_tensor_restride` which remove the need to compute
some contiguity flags at all.

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @aakhundov @kadeng

peterbell10 committed:
TensorImpl: Lazily compute numel and contiguity when symbolic

Currently whenever the sizes or strides are modified for a `TensorImpl` we
eagerly recompute the numel and memory format flags. This is fine for static
shapes as it's all fast C++ code, but for symbolic shapes it runs slow python code.

This instead changes the `SymbolicShapeMeta` object to compute the derived
quantities lazily at the first request. This has the added benefit that we can
now pass assumptions in `empty_tensor_restride` which remove the need to compute
some contiguity flags at all.

[ghstack-poisoned]

lezcano reviewed (commented):
inb4 splitting this into 2 PRs, one with codemovement and a second one with actual changes would be great.

peterbell10 committed:
Update on "TensorImpl: Lazily compute numel and contiguity when symbolic"


Currently whenever the sizes or strides are modified for a `TensorImpl` we
eagerly recompute the numel and memory format flags. This is fine for static
shapes as it's all fast C++ code, but for symbolic shapes it runs slow python code.

This instead changes the `SymbolicShapeMeta` object to compute the derived
quantities lazily at the first request. This has the added benefit that we can
now pass assumptions in `empty_tensor_restride` which remove the need to compute
some contiguity flags at all.

[ghstack-poisoned]

peterbell10 committed:
Update on "TensorImpl: Lazily compute numel and contiguity when symbolic"


Currently whenever the sizes or strides are modified for a `TensorImpl` we
eagerly recompute the numel and memory format flags. This is fine for static
shapes as it's all fast C++ code, but for symbolic shapes it runs slow python code.

This instead changes the `SymbolicShapeMeta` object to compute the derived
quantities lazily at the first request. This has the added benefit that we can
now pass assumptions in `empty_tensor_restride` which remove the need to compute
some contiguity flags at all.

[ghstack-poisoned]

peterbell10 committed:
Update on "TensorImpl: Lazily compute numel and contiguity when symbolic"


Currently whenever the sizes or strides are modified for a `TensorImpl` we
eagerly recompute the numel and memory format flags. This is fine for static
shapes as it's all fast C++ code, but for symbolic shapes it runs slow python code.

This instead changes the `SymbolicShapeMeta` object to compute the derived
quantities lazily at the first request. This has the added benefit that we can
now pass assumptions in `empty_tensor_restride` which remove the need to compute
some contiguity flags at all.

[ghstack-poisoned]

peterbell10 committed:
Fix Lint on "TensorImpl: Lazily compute numel and contiguity when symbolic"


Currently whenever the sizes or strides are modified for a `TensorImpl` we
eagerly recompute the numel and memory format flags. This is fine for static
shapes as it's all fast C++ code, but for symbolic shapes it runs slow python code.

This instead changes the `SymbolicShapeMeta` object to compute the derived
quantities lazily at the first request. This has the added benefit that we can
now pass assumptions in `empty_tensor_restride` which remove the need to compute
some contiguity flags at all.

[ghstack-poisoned]

peterbell10 committed:
Update on "TensorImpl: Lazily compute numel and contiguity when symbolic"


Currently whenever the sizes or strides are modified for a `TensorImpl` we
eagerly recompute the numel and memory format flags. This is fine for static
shapes as it's all fast C++ code, but for symbolic shapes it runs slow python code.

This instead changes the `SymbolicShapeMeta` object to compute the derived
quantities lazily at the first request. This has the added benefit that we can
now pass assumptions in `empty_tensor_restride` which remove the need to compute
some contiguity flags at all.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

peterbell10 commented:
Ran the benchmark suit and seeing around 10% compile time improvement for dynamic shapes.


| Suite       | Master  (s)            | This PR (s) | Speedup |
|-------------|-----------------------|---------|---------|
| TorchBench  | 85                    | 79     | 1.08    |
| HuggingFace | 122                   | 106     | 1.15    |
| TIMM Models | 147                   | 138     | 1.07    |
| [dynamic]   | 43                    | 29      | 1.48    |

Full results here:
https://hud.pytorch.org/benchmark/compilers?startTime=Tue%2C%2031%20Oct%202023%2011%3A40%3A21%20GMT&stopTime=Tue%2C%2007%20Nov%202023%2011%3A40%3A21%20GMT&granularity=hour&suite=torchbench&mode=training&dtype=amp&lBranch=gh/peterbell10/656/orig&lCommit=e236834d4186ce09934bee43ff28b1683d673757&rBranch=main&rCommit=2bc1378d7be563fa9b3050bb0e0fefd6e55a9e81

ezyang commented:
This isn't thread safe.

It should be possible to make it thread safe with minimal impact, as we eventually will call into Python which will be GIL protected. However, I don't see a clear recipe for doing so. I suppose we could allocate an atomic per field (or maybe only a coarse grained atomic is good enough?)

peterbell10 committed:
Thread safetly on "TensorImpl: Lazily compute numel and contiguity when symbolic"


Currently whenever the sizes or strides are modified for a `TensorImpl` we
eagerly recompute the numel and memory format flags. This is fine for static
shapes as it's all fast C++ code, but for symbolic shapes it runs slow python code.

This instead changes the `SymbolicShapeMeta` object to compute the derived
quantities lazily at the first request. This has the added benefit that we can
now pass assumptions in `empty_tensor_restride` which remove the need to compute
some contiguity flags at all.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

peterbell10 committed:
Update on "TensorImpl: Lazily compute numel and contiguity when symbolic"


Currently whenever the sizes or strides are modified for a `TensorImpl` we
eagerly recompute the numel and memory format flags. This is fine for static
shapes as it's all fast C++ code, but for symbolic shapes it runs slow python code.

This instead changes the `SymbolicShapeMeta` object to compute the derived
quantities lazily at the first request. This has the added benefit that we can
now pass assumptions in `empty_tensor_restride` which remove the need to compute
some contiguity flags at all.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

peterbell10 commented on c10/core/SymbolicShapeMeta.cpp:
```diff
@@ -114,57 +130,188 @@ static bool definitely_true(const SymBool& b) {
+    return;
+  }
+  numel_ = std::move(val);
+  available_.fetch_or(numel_avail);
```
Okay, here is a thread safety strategy. I use double-checked locking where the available bit in an atomic int indicates the variable is ready to read, and the write to the variable is protected by a mutex. The actual values are computed outside of the mutex lock which creates a small risk of calculating it twice, but the SymInt calculations cannot be done under the mutex since it might call arbitrary python code.

ezyang replied:
Recompute is fine, nice trick.

peterbell10 committed:
Fix lint on "TensorImpl: Lazily compute numel and contiguity when symbolic"


Currently whenever the sizes or strides are modified for a `TensorImpl` we
eagerly recompute the numel and memory format flags. This is fine for static
shapes as it's all fast C++ code, but for symbolic shapes it runs slow python code.

This instead changes the `SymbolicShapeMeta` object to compute the derived
quantities lazily at the first request. This has the added benefit that we can
now pass assumptions in `empty_tensor_restride` which remove the need to compute
some contiguity flags at all.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

peterbell10 committed:
Update on "TensorImpl: Lazily compute numel and contiguity when symbolic"


Currently whenever the sizes or strides are modified for a `TensorImpl` we
eagerly recompute the numel and memory format flags. This is fine for static
shapes as it's all fast C++ code, but for symbolic shapes it runs slow python code.

This instead changes the `SymbolicShapeMeta` object to compute the derived
quantities lazily at the first request. This has the added benefit that we can
now pass assumptions in `empty_tensor_restride` which remove the need to compute
some contiguity flags at all.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

peterbell10 committed:
More lint on "TensorImpl: Lazily compute numel and contiguity when symbolic"


Currently whenever the sizes or strides are modified for a `TensorImpl` we
eagerly recompute the numel and memory format flags. This is fine for static
shapes as it's all fast C++ code, but for symbolic shapes it runs slow python code.

This instead changes the `SymbolicShapeMeta` object to compute the derived
quantities lazily at the first request. This has the added benefit that we can
now pass assumptions in `empty_tensor_restride` which remove the need to compute
some contiguity flags at all.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

peterbell10 commented:
@pytorchbot merge

Flamefire commented:
While working on #117033 I found this PR and I'm a bit worried about using that mutex in `SymbolicShapeMeta` only in const methods. Is it impossible for 2 threads having references to the same instance and one accessing it non-const while the other reads it (i.e. as const)? That would have only the "other" lock it which means the data is still concurrently accessed.   
Second: The `available_.fetch_and(~numel_avail); numel_ = 1;`-pattern used in 2 places looks odd: If the `available`-set already means, `numel_` is invalid and needs to be recomputed on access, why set it? Can't now a race with `set_numel` happen, that does `numel_ = std::move(val); available_.fetch_or(numel_avail);`? I.e. one thread clears the flag, the other sets it and right after the first sets `numel_ =1` making the value be "available" but wrong.

peterbell10 commented:
> Is it impossible for 2 threads having references to the same instance and one accessing it non-const while the other reads it (i.e. as const)? 

That would indicate a bug as the mutex isn't intended to protect all concurrent access but only to provide the default expectation that concurrent const accesses won't race. This is a in line with the rest of `TensorImpl`.

> The `available_.fetch_and(~numel_avail); numel_ = 1;` -pattern used in 2 places looks odd: [...] Can't now a race with set_numel happen,

As before, `refresh_numel` is non-const so there is no expectation of threading guaruntees.

> If the `available`-set already means, `numel_` is invalid and needs to be recomputed on access, why set it?

numel may be a symbolic expression in python, so setting it releases that reference.
